### DISCUSSION TITLE
What is the expected format of rewards?

### DISCUSSION THREAD

**Discussion by the-big-bad-wolf** [Toni-SM/skrl#280] [Q&A] [answered]
I am trying to create a wrapper for the Waymax environment as mentioned [here](https://github.com/Toni-SM/skrl/discussions/275). But I am having trouble understanding how the reward is supposed to be formatted. Unless I wrap the reward i return in my custom step function twice my code crashes due to indexing errors at line 372 in `skrl.agents.jax.base.py` in the `record_transition` function:

` self._track_rewards.extend(self._cumulative_rewards[finished_episodes][:, 0].reshape(-1).tolist())
`

As I understand it `finished_episodes` is to index on the environment if you are running multiple environments in parallel. But I am not sure what the next two indices are for.

**Comment by Toni-SM** [comment, marked as answer]
Hi @the-big-bad-wolf 

The data expected/returned by the wrappers (except the `info`, a dictionary) must be **flattened tensors/arrays** with shape `(N, data-size)`, where N is the number of environments (one or many).

| Data         | Shape           | Dtype   |
|--------------|-----------------|---------|
| actions | `(N, action_space size)` | `float32` |
| observations | `(N, observation_space size)` | `float32` |
| states | `(N, state_space size)` | `float32` |
| rewards      | `(N, 1)`          | `float32` |
| terminated   | `(N, 1)`          | `bool`    |
| truncated    | `(N, 1)`          | `bool`    |
| info (dict)        | -                | -         |

Use the skrl's [spaces utilities](https://skrl.readthedocs.io/en/latest/api/utils/spaces.html) to transform and operate with the spaces

**Reply by the-big-bad-wolf** [reply]
I now have the correct shapes, but I run into a problem when using jax as the backend instead of numpy. When using jax as the backend the training seems to stop progressing after around 10 iterations. If I change the backend to numpy and return numpy arrays in the step and reset functions the problem disappears and training finishes as expected, so I am not sure what the problem is. For context each episode of the environment lasts 80 steps, so it isn't due to a problem in the reset. CPU usage remains high during this time.

You can see the implementation of the Wrapper at https://github.com/the-big-bad-wolf/WaymaxAutoPlanner/blob/main/main.py

**Reply by Toni-SM** [reply]
Hi @the-big-bad-wolf 

Does this last problem still occur with the `"jax"` backend?

**Reply by the-big-bad-wolf** [reply]
Turns out it didn't stop, it just took a long time, so I thought it had stalled. It is probably just the JIT compilation time as you mentioned [here](https://github.com/Toni-SM/skrl/discussions/286#discussioncomment-12475379). I have switched to using numpy in the backend anyway.
